lib/src/authentication/token_store.dart

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// ignore_for_file: import_of_legacy_library_into_null_safe
+
+import 'dart:convert';
+
+import 'package:path/path.dart' as path;
+
+import '../io.dart';
+import '../log.dart' as log;
+import '../source/hosted.dart';
+import 'token.dart';
+
+/// Stores and manages authentication credentials.
+class TokenStore {
+  TokenStore(this.cacheRootDir);
+
+  /// Cache directory.
+  final String cacheRootDir;
+
+  List<Token>? _tokens;
+
+  /// List of saved authentication tokens.
+  ///
+  /// Modifying this field will not write changes to the disk. You have to call
+  /// [flush] to save changes.
+  List<Token> get tokens => _tokens ??= _loadTokens();
+
+  /// Reads "tokens.json" and parses / deserializes it into list of
+  /// [Token].
+  List<Token> _loadTokens() {
+    final result = List<Token>.empty(growable: true);
+    final path = _tokensFile;
+    if (!fileExists(path)) {
+      return result;
+    }
+
+    try {
+      final json = jsonDecode(readTextFile(path));
+
+      if (json is! Map<String, dynamic>) {
+        throw FormatException('JSON contents is corrupted or not supported.');
+      }
+      if (json['version'] != 1) {
+        throw FormatException('Version is not supported.');
+      }
+
+      if (json.containsKey('hosted')) {
+        if (json['hosted'] is! List) {
+          throw FormatException(
+              'tokens.json format is invalid or not supported.');
+        }
+
+        result.addAll((json['hosted'] as List)
+            .cast<Map<String, dynamic>>()
+            .map((it) => Token.fromJson(it)));
+      }
+    } on FormatException catch (error, stackTrace) {
+      log.error('Failed to load tokens.json', error, stackTrace);
+
+      // When an invalid, damaged or not compatible version of token.json is
+      // found, we remove it after showing error message. Otherwise the error
+      // message will be displayed on each pub command.
+      // Or instead we could write instructrions to execute
+      //`pub token remove --all` if user couldn't solve the issue.
+      deleteEntry(_tokensFile);
```

Comment:
yes, let's not delete the file.

Then it'll be overwritten when  the user does: `dart pub token add`, and user will get a warning in log saying that it failed to load tokens.json